neo4j/internal/homedb/cache.go

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [https://neo4j.com]
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package homedb
+
+import (
+	"fmt"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/auth"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/errorutil"
+	"sort"
+	"sync"
+	"time"
+)
+
+// DefaultCacheMaxSize defines the maximum number of entries the cache can hold.
+const DefaultCacheMaxSize = 1000
+
+// cachePruneFraction defines the fraction of entries to prune when the cache exceeds its size.
+const cachePruneFraction = 0.1
+
+const keyScheme = "scheme"
+const schemeNone = "none"
+const schemeBasic = "basic"
+const schemeKerberos = "kerberos"
+const schemeBearer = "bearer"
+const keyPrincipal = "principal"
+const keyCredentials = "credentials"
+const keyRealm = "realm"
+const keyParameters = "parameters"
+
+type cacheEntry struct {
+	database string
+	lastUsed time.Time
+}
+
+type Cache struct {
+	maxSize int
+	cache   map[string]*cacheEntry
+	enabled bool
+	mu      sync.RWMutex
+}
+
+// NewCache creates and returns a new cache instance with the given max size.
+func NewCache(maxSize int) (*Cache, error) {
+	if maxSize <= 0 {
+		return nil, &errorutil.UsageError{Message: "Maximum cache size must be greater than 0"}
+	}
+	return &Cache{
+		maxSize: maxSize,
+		cache:   make(map[string]*cacheEntry),
```

Comment:
```suggestion
		cache:   make(map[string]*cacheEntry, maxSize + 1),
```
Not sure if worth it... Saves some potential allocations later down the road at the cost of a higher memory footprint (but probably not by much).

Reply:
probs worth the footprint, updated 👍